packages/collision/index.ts

```diff
@@ -1,111 +1,449 @@
+      arr.push(f);
+    });
+  }
+}
+
+function prepareProxyFunction<T extends Function>(functions: Function[], itemName: string, proxyFn: T): T {
+  defineProperty(proxyFn, 'name', { value: itemName, configurable: true });
+  defineProperty(proxyFn, AGGREGATION_PROPERTY_NAME, { value: functions, configurable: true });
+  return proxyFn;
+}
+
+function makeMapSyncProxyFunction({ functions, itemName }: MakeProxyFunctionOptions): MapSyncProxyFunction {
+  return prepareProxyFunction(functions, itemName, function mapSyncFn(this: object, ...args: unknown[]): unknown[] {
+    const fns = getDomainItemAggregates(mapSyncFn);
+    return [...fns.map((fn) => fn.apply(this, [...args]))];
+  });
+}
+
+function makeReduceSyncProxyFunction({ functions, itemName }: MakeProxyFunctionOptions): ReduceSyncProxyFunction {
+  return prepareProxyFunction(
+    functions,
+    itemName,
+    function reduceSyncFn(initialValue: unknown, ...args: unknown[]): unknown {
+      return getDomainItemAggregates(reduceSyncFn).reduce((o, fn) => {
+        return fn(o, ...args);
+      }, initialValue);
+    }
+  );
+}
+
+function makeReduceThisSyncProxyFunction({ functions, itemName }: MakeProxyFunctionOptions): ReduceSyncProxyFunction {
+  return prepareProxyFunction(
+    functions,
+    itemName,
+    function reduceThisSyncFn(this: object, ...args: unknown[]): unknown {
+      return getDomainItemAggregates(reduceThisSyncFn).reduce((o, fn) => {
+        const result = fn.apply(o, args);
+        return result || this;
+      }, this);
+    }
+  );
+}
+
+function makeMapAsyncProxyFunction({ functions, itemName }: MakeProxyFunctionOptions): MapAsyncProxyFunction {
+  return prepareProxyFunction(functions, itemName, async function mapAsyncFn(this: object, ...args: unknown[]): Promise<
+    unknown[]
+  > {
+    return Promise.all(getDomainItemAggregates(mapAsyncFn).map((fn) => fn.apply(this, args)));
+  });
+}
+
+function makeReduceAsyncProxyFunction({ functions, itemName }: MakeProxyFunctionOptions): ReduceAsyncProxyFunction {
+  return prepareProxyFunction(
+    functions,
+    itemName,
+    async function reduceAsyncFn(initialValue: unknown, ...args: unknown[]): Promise<unknown> {
+      return getDomainItemAggregates(reduceAsyncFn).reduce((promise, fn) => {
+        return fn(promise, ...args);
+      }, Promise.resolve(initialValue));
+    }
+  );
+}
+
+function makeReduceThisAsyncProxyFunction({ functions, itemName }: MakeProxyFunctionOptions): ReduceAsyncProxyFunction {
+  return prepareProxyFunction(
+    functions,
+    itemName,
+    async function reduceThisAsyncFn(this: object, ...args: unknown[]): Promise<unknown> {
+      return getDomainItemAggregates(reduceThisAsyncFn).reduce((promise, fn) => {
+        return promise.then((result) => {
+          return Promise.resolve(fn.apply(result || this, args)).then((nextResult) => {
+            return nextResult || this;
+          });
+        });
+      }, Promise.resolve(this || this));
```

Comment:
`this || this` ? :)))

Reply:
Is that really in there? Lol. Probably a VS Code symbol rename victim. Will check it out.

Regarding the published vs repo thing... We have to start somewhere. I'll help @PopGoesTheWza with merging it through